### PR TITLE
REF: use new pytables helper methods to de-duplicate _convert_index

### PR DESCRIPTION
**Pull request by jbrockmendel**
The assertions in here are intended to be temporary.  Once reviewers are convinced this is accurate, then the assertions can be removed and a bunch of these blocks can be condensed down to share code.

The new helper func _get_data_and_dtype_name is also going to be used later to pre-empt the need for the state-altering `get_data` method.

### CODE REVIEW

File: pandas/io/pytables.py [hunk 1]
```diff
@@ -2318,6 +2309,9 @@ def get_atom_coltype(cls, kind: str) -> Type["Col"]:
         if kind.startswith("uint"):
             k4 = kind[4:]
             col_name = f"UInt{k4}Col"
+        elif kind.startswith("period"):
+            # we store as integer
```
Comment (jreback) [inline pandas/io/pytables.py]: we may have very little coverage for PeriodIndex FYI

File: pandas/io/pytables.py [hunk 2]
```diff
@@ -4924,21 +4921,47 @@ def _dtype_to_kind(dtype_str: str) -> str:
...
+    """
+    Convert the passed data into a storable form and a dtype string.
+    """
+    if is_categorical_dtype(data.dtype):
+        data = data.codes
+
+    # For datetime64tz we need to drop the TZ in tests TODO: why?
+    dtype_name = data.dtype.name.split("[")[0]
+
+    if data.dtype.kind in ["m", "M"]:
```
Comment (jreback) [inline pandas/io/pytables.py]: use needs_i8_conversion (followon ok)

File: pandas/io/pytables.py [hunk 3]
```diff
@@ -4924,21 +4921,47 @@ def _dtype_to_kind(dtype_str: str) -> str:
...
+        data = np.asarray(data.view("i8"))
+        # TODO: we used to reshape for the dt64tz case, but no longer
+        #  doing that doesnt seem to break anything.  why?
+
+    elif isinstance(data, PeriodIndex):
```
Comment (jreback) [inline pandas/io/pytables.py]: can likely use extract_array here

**Review by jreback — CHANGES_REQUESTED** [review]
looks good about the assertions

File: pandas/io/pytables.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/io/pytables.py]: that would give PeriodArray, we need to get i8 back here

Comment (jreback) [reply pandas/io/pytables.py]: right once you have the arrays, then you can simply .view('i8') and you are done (pretty generically)

Comment (jbrockmendel) [reply pandas/io/pytables.py]: ok i see what you're saying now.  will do on next pass (along with needs_i8 mentioned above)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks good. if you can do follows as indicated above when you have a chance.

**Pull request merged**